File: adapters/codex/hooks/packages/agent-state/filelock.js
'use strict';

const fs = require('fs');

// Lock protocol parameters MUST match internal/state/filelock.go so Go and
// Node writers on the same agent file rendezvous on the same sidecar.
const LOCK_SUFFIX = '.lock';
const LOCK_STALE_MS = 5000;
const LOCK_RETRY_MS = 5;
const LOCK_MAX_RETRIES = 400; // ~2s budget

function sleepSync(ms) {
  // Atomics.wait is the only built-in synchronous sleep in Node — no spawn,
  // no busy-wait. writeState is synchronous so we cannot await/setTimeout.
  const sab = new SharedArrayBuffer(4);
  const view = new Int32Array(sab);
  Atomics.wait(view, 0, 0, ms);
}

/**
 * Run fn while holding an exclusive sidecar lock on filePath.
 * Created with O_CREAT|O_EXCL — the same atomic primitive used by Go's
 * withFileLock, so cross-language writers coordinate.
 *
 * A lock older than LOCK_STALE_MS is treated as abandoned (writer crashed)
 * and removed. After LOCK_MAX_RETRIES the call returns the lock-acquire
 * error to the caller rather than blocking indefinitely.
 *
 * @param {string} filePath - the data file being protected
 * @param {function} fn - synchronous critical section
 * @returns whatever fn returns
 */
function withFileLock(filePath, fn) {
  const lockPath = filePath + LOCK_SUFFIX;
  let fd = -1;
  for (let attempt = 0; attempt < LOCK_MAX_RETRIES; attempt++) {
    try {
      fd = fs.openSync(lockPath, 'wx', 0o600);
      break;
    } catch (err) {
      if (err.code !== 'EEXIST') {
        throw err;
      }
      try {
        const stat = fs.statSync(lockPath);
        if (Date.now() - stat.mtimeMs > LOCK_STALE_MS) {
          try { fs.unlinkSync(lockPath); } catch { /* race with another acquirer */ }
          continue;
        }
      } catch { /* lock vanished between EEXIST and stat — retry */ }
      sleepSync(LOCK_RETRY_MS);
    }
  }
  if (fd === -1) {
    throw new Error(`acquire lock for ${filePath}: timed out`);
  }
  try {
    try { fs.writeSync(fd, String(process.pid)); } catch { /* best-effort */ }
    fs.closeSync(fd);
    return fn();
  } finally {
    try { fs.unlinkSync(lockPath); } catch { /* already removed */ }
  }
}

module.exports = { withFileLock };

File: internal/state/filelock.go
package state

import (
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"time"
)

const (
	lockSuffix     = ".lock"
	lockStaleAfter = 5 * time.Second
	lockRetryWait  = 5 * time.Millisecond
	lockMaxRetries = 400
)

// withFileLock acquires an exclusive sidecar lock for filePath, runs fn,
// then releases the lock. The lock file is filePath + ".lock" and is
// created via O_CREATE|O_EXCL — the same atomic primitive the JS writer
// uses, so Go and Node writers on the same agent file rendezvous on the
// same kernel object.
//
// A lock older than lockStaleAfter is treated as abandoned (the writer
// crashed) and unlinked. Retry budget is bounded; if the lock cannot be
// acquired within ~2s, withFileLock returns an error rather than block
// the caller indefinitely.
func withFileLock(filePath string, fn func() error) error {
	lockPath := filePath + lockSuffix
	for attempt := 0; attempt < lockMaxRetries; attempt++ {
		f, err := os.OpenFile(lockPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0600)
		if err == nil {
			_, _ = fmt.Fprintf(f, "%d", os.Getpid())
			_ = f.Close()
			defer os.Remove(lockPath)
			return fn()
		}
		if !errors.Is(err, fs.ErrExist) {
			return fmt.Errorf("open lock %s: %w", lockPath, err)
		}
		if info, statErr := os.Stat(lockPath); statErr == nil {
			if time.Since(info.ModTime()) > lockStaleAfter {
				_ = os.Remove(lockPath)
				continue
			}
		}
		time.Sleep(lockRetryWait)
	}
	return fmt.Errorf("acquire lock for %s: timed out", filePath)
}

// writeJSONAtomic writes data to filePath via tmp + rename. The rename is
// atomic on POSIX, so a concurrent reader sees either the old contents or
// the new — never a truncated/empty file. Fixes the torn-read race where
// os.WriteFile's truncate-then-write window let JS hooks observe an empty
// file and merge against {} (wiping every existing field).
func writeJSONAtomic(filePath string, data []byte) error {
	dir := filepath.Dir(filePath)
	base := filepath.Base(filePath)
	tmp := filepath.Join(dir, fmt.Sprintf(".%s.tmp.%d", base, os.Getpid()))
	if err := os.WriteFile(tmp, data, 0600); err != nil {
		return fmt.Errorf("write tmp: %w", err)
	}
	if err := os.Rename(tmp, filePath); err != nil {
		_ = os.Remove(tmp)
		return fmt.Errorf("rename %s -> %s: %w", tmp, filePath, err)
	}
	return nil
}
